### ISSUE TITLE
EntryLoaderSimpleTest.testValues

### ISSUE THREAD

**Issue by vojtechtoman** [hazelcast/hazelcast#15677]
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3615

com.hazelcast.map.impl.mapstore.EntryLoaderSimpleTest.testValues[inMemoryFormat: BINARY]

```
expected null, but was:<val0>

Stacktrace

java.lang.AssertionError: expected null, but was:<val0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotNull(Assert.java:755)
	at org.junit.Assert.assertNull(Assert.java:737)
	at org.junit.Assert.assertNull(Assert.java:747)
	at com.hazelcast.map.impl.mapstore.EntryLoaderSimpleTest.lambda$assertExpiredEventually$0(EntryLoaderSimpleTest.java:339)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1316)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1418)
	at com.hazelcast.map.impl.mapstore.EntryLoaderSimpleTest.assertExpiredEventually(EntryLoaderSimpleTest.java:337)
	at com.hazelcast.map.impl.mapstore.EntryLoaderSimpleTest.testValues(EntryLoaderSimpleTest.java:287)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by vojtechtoman** [comment]
The test uses an `EntryLoader` to load records with a set expiration time (T = 10s from now) and then checks that the records expire eventually.
One potential case when a record created this way may not expire is when the record is created in`DefaultRecordStore.loadRecordOrNull` _exactly_ at time T. Since the expiry time provided by the entry loader is converted to a TTL internally, it can happen that the calculated TTL is 0 - in which case it is ignored and the record will be non-expirable. Having said that, I refuse to believe that this is what happened as such unlikely scenario would signify a clear flaw in the Matrix.

**Comment by tezc** [comment]
http://jenkins.hazelcast.com/job/Hazelcast-3.x-OracleJDK10/49/testReport/com.hazelcast.map.impl.mapstore/EntryLoaderSimpleTest/testKeySet_inMemoryFormat__OBJECT_/

**Comment by ihsandemir** [comment]
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3929/testReport/junit/com.hazelcast.map.impl.mapstore/EntryLoaderSimpleTest/testEntrySet_inMemoryFormat__BINARY_/

**Comment by Holmistr** [comment]
Logging enhancement was merged. Closing this issue and let's wait for another failure. Reopen if needed.

**Comment by vojtechtoman** [comment]
Finally, it is clear to me what is going on. It's a combination of how `EntryLoaderSimpleTest.assertExpiredEventually()` is implemented with my initial "impossible" scenario description above. I managed to reproduce it locally eventually. As often, the explanation is quite straightforward once all pieces of debug logging, timing etc. fall together.

In short, the the ttl and maxIdle of the `key0`/`val0` record was `LONG.MAX_LONG` even though the test sets the expiration to 10 seconds from now. Everything works as expected until we enter `assertExpiredEventually()`: the records are created with the correct expiration time, the test checks that the entries / values etc. are still in the map (with the correct expiration times). But the way `assertExpiredEventually()` is implemented, it simply loops over all records and performs `map.get()` until all return `null`. However, the problem with this approach is that once a record expires and `map.get()` returns `null`, the subsequent `map.get()` on the same key triggers load from the entry loader. Under most circumstances, the loader logic will see that the expiration time of the loaded entry is older than the current time and the entry is ignored. **However, if the loader happens to be called at the exact time (in millis) as the entry expiration time, the entry will be loaded with ttl and maxIdle equal to Long.MAX_IDLE (the "impossible" scenario above).**

As a lucky coincidence, I believe my commit e3deee49ba5c258dd25b7be0e7c9ae0ca6101a31 (in which I rewrote `assertExpiredEventually()` to use `map.getEntryView()` instead of `map.get()` to get more debug info in the logs) actually fixes the tests. It does not address the "impossible" issue itself (I will create a separate issue for that one), but it prevents it from occurring.
